src/Storages/StorageMerge.cpp

```diff
@@ -616,22 +592,109 @@ SelectQueryInfo ReadFromMerge::getModifiedQueryInfo(const SelectQueryInfo & quer
+
+    auto modified_query_info = getModifiedQueryInfo(query_info, context, storage_with_lock, nested_storage_snaphsot);
+    Names column_names_as_aliases;
+
+    if (!context->getSettingsRef().allow_experimental_analyzer)
+    {
+        auto storage_columns = storage_metadata_snapshot->getColumns();
+        auto syntax_result = TreeRewriter(context).analyzeSelect(
+            modified_query_info.query, TreeRewriterResult({}, storage, nested_storage_snaphsot));
+
+        bool with_aliases = common_processed_stage == QueryProcessingStage::FetchColumns && !storage_columns.getAliases().empty();
+        if (with_aliases)
+        {
+            ASTPtr required_columns_expr_list = std::make_shared<ASTExpressionList>();
+            ASTPtr column_expr;
+
+            for (const auto & column : real_column_names)
+            {
+                const auto column_default = storage_columns.getDefault(column);
+                bool is_alias = column_default && column_default->kind == ColumnDefaultKind::Alias;
+
+                if (is_alias)
+                {
+                    column_expr = column_default->expression->clone();
+                    replaceAliasColumnsInQuery(column_expr, storage_metadata_snapshot->getColumns(),
+                        syntax_result->array_join_result_to_source, context);
+
+                    const auto & column_description = storage_columns.get(column);
+                    column_expr = addTypeConversionToAST(std::move(column_expr), column_description.type->getName(),
+                        storage_metadata_snapshot->getColumns().getAll(), context);
+                    column_expr = setAlias(column_expr, column);
+
+                    auto type = sample_block.has(column) ? sample_block.getByName(column).type : column_description.type;
+
+                    aliases.push_back({ .name = column, .type = type, .expression = column_expr->clone() });
+                }
+                else
+                    column_expr = std::make_shared<ASTIdentifier>(column);
+
+                required_columns_expr_list->children.emplace_back(std::move(column_expr));
+            }
+
+            syntax_result = TreeRewriter(context).analyze(
+                required_columns_expr_list, storage_columns.getAllPhysical(), storage, storage->getStorageSnapshot(storage_metadata_snapshot, context));
+
+            auto alias_actions = ExpressionAnalyzer(required_columns_expr_list, syntax_result, context).getActionsDAG(true);
+
+            column_names_as_aliases = alias_actions->getRequiredColumns().getNames();
+
+            if (column_names_as_aliases.empty())
+                column_names_as_aliases.push_back(ExpressionActions::getSmallestColumn(storage_metadata_snapshot->getColumns().getAllPhysical()).name);
+        }
+    }
+    if (!column_names_as_aliases.empty())
+    {
+        real_column_names = column_names_as_aliases;
```

Comment:
So it replaces `real_column_names` only if `!column_names_as_aliases.empty()`
Such a complicated logic.

Reply:
Let's make the function `processAliases()` return `column_names_as_aliases` as is, and do replace `read_column_names` not here but in the function which calls `processAliases()`

Reply:
You've made an interference already, it doesn't add too much interference if we pass arguments this way or another. I'm ok with passing results via reference parameters. For me it's just better to avoid parameters which are input and output at the same time - just to make the life easier. So why can't we make `column_names_as_aliases` a pure output parameter and do this logic in `createSource()`:

```
... createSource(...)
{
    ...
    Aliases aliases;
    Names column_names_as_aliases;
    processAliases(context, storage_with_lock, sample_block, modified_context, aliases, column_names_as_aliases);
    if (!column_names_as_aliases.empty())
        real_column_names = column_names_as_aliases;
    ...
```

Reply:
May be because this logic is a part of processing aliases ? ;)